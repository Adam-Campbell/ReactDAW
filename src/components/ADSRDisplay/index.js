import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Stage, Layer, Rect, Line } from 'react-konva';
import { UIColors } from '../../constants';

/*

A, D, R => width ranges from 0 to 60, height always 100
S => width always 60, height ranges from 0 to 100.
Line defined by a points array expressed as [x1, y1, x2, y2]

To work out the horizontal position for A, D, R, take the current value x, and (x / 0.995) * 60
To work out vertical position for S, 100 - ((x / 0.995) * 100)

Defining the line points.

let ah = horizontal position for a
let dh = horizontal position for d
let sv = vertical position for s
let rh = horizontal position for release

points for a = [0, 0, ah, 100]
points for d = [ah, 100, ah+dh, sv]
points for s = [ah+dh, sv, ah+dh+60, sv]
points for r = [ah+dh+60, sv, ah+dh+60+rg, 0]

*/

export const calculateLinePoints = (attack, decay, sustain, release) => {
    const aW = (attack / 0.995) * 60;
    const dW = (decay / 0.995) * 60;
    const sH = 100 - ((sustain / 0.995) * 100);
    const rW = (release / 0.995) * 60; 
    return {
        attackPoints: [0, 100, aW, 0],
        decayPoints: [aW, 0, aW+dW, sH],
        sustainPoints: [aW+dW, sH, aW+dW+60, sH],
        releasePoints: [aW+dW+60, sH, aW+dW+60+rW, 100]
    };
}

class ADSRDisplay extends Component {

    static propTypes = {
        attack: PropTypes.number.isRequired,
        decay: PropTypes.number.isRequired,
        sustain: PropTypes.number.isRequired,
        release: PropTypes.number.isRequired,
        id: PropTypes.string.isRequired
    }

    constructor(props) {
        super(props);
        this.stageRef = React.createRef();
    }

    componentWillUnmount() {
        if (this.stageRef.current) {
            this.stageRef.current.content = null;
        }
    }

    render() {
        const {
            attackPoints,
            decayPoints,
            sustainPoints,
            releasePoints
        } = calculateLinePoints(
            this.props.attack, 
            this.props.decay, 
            this.props.sustain, 
            this.props.release
        );
        return (
            <div id={this.props.id}>
                <Stage
                    container={this.props.id}
                    ref={this.stageRef}
                    width={300} 
                    height={120} 
                >
                    <Layer>
                        <Rect 
                            x={0}
                            y={0}
                            width={300}
                            height={120}
                            fill={UIColors.lightPurple}
                        />
                    </Layer>
                    <Layer x={30} y={10}>
                        <Line 
                            points={attackPoints}
                            listening={false}
                            stroke={UIColors.offWhite}
                            strokeWidth={3}  
                        />
                        <Line 
                            points={decayPoints}
                            listening={false}
                            stroke={UIColors.offWhite}
                            strokeWidth={3}  
                        />
                        <Line 
                            points={sustainPoints}
                            listening={false}
                            stroke={UIColors.offWhite}
                            strokeWidth={3}  
                        />
                        <Line 
                            points={releasePoints}
                            listening={false}
                            stroke={UIColors.offWhite}
                            strokeWidth={3}  
                        />
                    </Layer>
                </Stage>
            </div>
        )
    }
}

export default ADSRDisplay;