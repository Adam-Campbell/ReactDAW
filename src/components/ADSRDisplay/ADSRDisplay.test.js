import React from 'react';
import ADSRDisplay, { calculateLinePoints } from './';
import { Stage, Layer, Rect, Line } from 'react-konva';
import { shallow } from 'enzyme';

describe('calculateLinePoints', () => {
    test('correctly calculates line points based on the inputs given', () => {
        expect(calculateLinePoints(0.995, 0.4975, 0.4975, 0.995))
        .toEqual({
            attackPoints: [0, 100, 60, 0],
            decayPoints: [60, 0, 90, 50],
            sustainPoints: [90, 50, 150, 50],
            releasePoints: [150, 50, 210, 100]
        });
    });
});

describe('ADSRDisplay', () => {
    const wrapper = shallow(
        <ADSRDisplay 
            attack={0.995}
            decay={0.4975}
            sustain={0.4975}
            release={0.995}
            id="adsr-display"
        />
    );
    test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('passes supplied id prop to the correct components', () => {
        expect(wrapper.find('div#adsr-display')).toHaveLength(1);
        expect(wrapper.find(Stage).first().prop('container')).toBe('adsr-display');
    });
    test('renders the correct lines for the props supplied', () => {
        const linesWrapper = wrapper.find(Line);
        expect(linesWrapper.at(0).shallow().prop('points')).toEqual([0, 100, 60, 0]);
        expect(linesWrapper.at(1).shallow().prop('points')).toEqual([60, 0, 90, 50]);
        expect(linesWrapper.at(2).shallow().prop('points')).toEqual([90, 50, 150, 50]);
        expect(linesWrapper.at(3).shallow().prop('points')).toEqual([150, 50, 210, 100]);
    });
});