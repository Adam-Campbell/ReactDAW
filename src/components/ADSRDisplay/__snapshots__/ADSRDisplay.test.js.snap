// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ADSRDisplay renders correctly 1`] = `
<div
  id="adsr-display"
>
  <ForwardRef
    container="adsr-display"
    height={120}
    width={300}
  >
    <Layer>
      <Rect
        fill="#252f48"
        height={120}
        width={300}
        x={0}
        y={0}
      />
    </Layer>
    <Layer
      x={30}
      y={10}
    >
      <Line
        listening={false}
        points={
          Array [
            0,
            100,
            60,
            0,
          ]
        }
        stroke="#e0e0e0"
        strokeWidth={3}
      />
      <Line
        listening={false}
        points={
          Array [
            60,
            0,
            90,
            50,
          ]
        }
        stroke="#e0e0e0"
        strokeWidth={3}
      />
      <Line
        listening={false}
        points={
          Array [
            90,
            50,
            150,
            50,
          ]
        }
        stroke="#e0e0e0"
        strokeWidth={3}
      />
      <Line
        listening={false}
        points={
          Array [
            150,
            50,
            210,
            100,
          ]
        }
        stroke="#e0e0e0"
        strokeWidth={3}
      />
    </Layer>
  </ForwardRef>
</div>
`;
